lightning/src/offers/signer.rs

```diff
@@ -132,20 +148,33 @@ impl PartialEq for Metadata {
 pub(super) struct MetadataMaterial {
 	nonce: Nonce,
 	hmac: HmacEngine<Sha256>,
+	// Some for payer metadata and None for offer metadata
+	encrypted_payment_id: Option<PaymentId>,
 }
 
 impl MetadataMaterial {
-	pub fn new(nonce: Nonce, expanded_key: &ExpandedKey, iv_bytes: &[u8; IV_LEN]) -> Self {
+	pub fn new(
+		nonce: Nonce, expanded_key: &ExpandedKey, iv_bytes: &[u8; IV_LEN],
+		payment_id: Option<PaymentId>
+	) -> Self {
+		// Encrypt payment_id
+		let encrypted_payment_id = payment_id.map(|payment_id| {
+			PaymentId(expanded_key.crypt_for_offer(payment_id.0, &nonce.0))
```

Comment:
I think it's not the best practice to reuse nonces, not sure if it matters here though.

Reply:
@TheBlueMatt Would like your expertise here.

Reply:
So there's actually two different uses here.
- For `crypt_for_offer`, the nonce is used as the IV for `ChaCha20::get_single_block` to encrypt the payment id.
- For the `hmac_for_offer`, the nonce is used as one input along with an IV specific to the message type.

For the latter, `WITH[OUT]_ENCRYPTED_PAYMENT_ID_HMAC_INPUT` are also provided though it seems to me those might not be strictly necessary given the IV for the message type dictates whether a payment id is included in the input, in practice.

Does that sound right?